community/ssl/src/main/java/org/neo4j/ssl/SslPolicy.java

```diff
@@ -101,22 +115,63 @@ public SslContext nettyClientContext() throws SSLException
 
     public SslHandler nettyServerHandler( Channel channel ) throws SSLException
     {
-        return makeNettyHandler( channel, nettyServerContext() );
+        return makeNettyHandler( channel, nettyServerContext(), SslHandler::new );
     }
 
     public SslHandler nettyClientHandler( Channel channel ) throws SSLException
     {
-        return makeNettyHandler( channel, nettyClientContext() );
-    }
-
-    private SslHandler makeNettyHandler( Channel channel, SslContext sslContext )
+        return makeNettyHandler( channel, nettyClientContext(), sslEngine -> new SslHandler( sslEngine )
```

Comment:
Can this manual bootstrapping of SNI be replaced with:

```java
InetSocketAddress address = (InetSocketAddress) channel.remoteAddress();
SSLEngine sslEngine = sslContext.newEngine(channel.alloc(), address.getHostName(), address.getPort());
```

instead of

```java
SSLEngine sslEngine = sslContext.newEngine( channel.alloc() );
```

in `#makeNettyHandler(Channel, SslContext, Function<SSLEngine,SslHandler>)`?

Reply:
`channel.remoteAddress()` is not set at this point in time - it's set after the connection is established and we have to wire up SSL before the connection is established.

Reply:
Fair enough. Is it maybe possible to somehow get hold of the address we are connecting to in this method?
Presumably, there exists an `org.neo4j.helpers.SocketAddress` somewhere higher in the call stack with known hostname and port. It could be easier to pass that around instead of doing manual bootstrapping of SNI. Just an idea.

Reply:
This is the best I could come up with without significant refactoring.